internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -135,155 +130,117 @@ func (b *batcher) genBucketID(node *commonpb.Node, resource *resourcepb.Resource
-		} else {
-			stats.RecordWithTags(context.Background(), statsTags, statBatchOnDeadNode.M(1))
-		}
-		// Shoot off event so that we don't block this add on other adds
-		go nb.cutBatch(b)
-	}
-}
+func (nb *nodeBatch) add(spans []*tracepb.Span) {
+	nb.mu.Lock()
+	nb.items = append(nb.items, spans)
+	nb.totalItemCount = nb.totalItemCount + uint32(len(spans))
+	nb.cyclesUntouched = 0
 
-// cutBatch cuts the provided batch, and sets a new batch on this nodeBatcher
-func (nb *nodeBatcher) cutBatch(b *batch) {
-	initialCap := b.getCurrentCap()
-	currSize := b.getCurrentItemCount()
-	reducedCap := initialCap >> 1
-	for currSize < reducedCap && reducedCap > initialBatchCapacity {
-		initialCap = reducedCap
-		reducedCap = reducedCap >> 1
+	itemCount := nb.totalItemCount
+	var itemsToProcess [][]*tracepb.Span
+	if nb.totalItemCount > nb.parent.sendBatchSize || nb.dead == nodeStatusDead {
+		itemsToProcess, itemCount = nb.getAndReset()
 	}
-	// atomic.CompareAndSwapPointer only takes unsafe.Pointer interfaces. We do not use unsafe
-	// to skirt around the golang type system.
-	currBatchPtr := (*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))
-	swapped := atomic.CompareAndSwapPointer(
-		currBatchPtr,
-		unsafe.Pointer(b),
-		unsafe.Pointer(newBatch(initialCap, nb.sendBatchSize)),
-	)
-	// Since we are doing an atomic compare and swap, this batch will only be sent once.
-	if swapped {
-		b.closeBatch()
-		atomic.StoreInt64(&nb.lastSent, time.Now().UnixNano())
-		nb.sendBatch(b)
+	nb.mu.Unlock()
+
+	if len(itemsToProcess) > 0 {
+
+		nb.sendItems(itemsToProcess, itemCount, statBatchSizeTriggerSend)
 	}
 }
 
-func (nb *nodeBatcher) sendBatch(batch *batch) {
-	spans := batch.getSpans()
-	statsTags := processor.StatsTagsForBatch(
-		nb.parent.name, processor.ServiceNameForNode(nb.node), nb.spanFormat,
-	)
-	stats.RecordWithTags(context.Background(), statsTags, statBatchSize.M(int64(len(spans))))
-
-	if len(spans) == 0 {
-		return
+func (nb *nodeBatch) sendItems(
+	itemsToProcess [][]*tracepb.Span,
+	itemCount uint32,
+	measure *stats.Int64Measure,
+) {
+	tdItems := make([]*tracepb.Span, 0, itemCount)
+	for _, items := range itemsToProcess {
+		tdItems = append(tdItems, items...)
 	}
 	td := data.TraceData{
 		Node:     nb.node,
 		Resource: nb.resource,
-		Spans:    spans,
-	}
-	err := nb.parent.sender.ProcessSpans(td, nb.spanFormat)
-	// Assumed that the next processor always handles a batch, and doesn't error
-	if err != nil {
-		nb.logger.Error(
-			"Failed to process batch, discarding",
-			zap.String("processor", "batcher"),
-			zap.Uint32("batch-size", batch.getCurrentItemCount()),
-		)
+		Spans:    tdItems,
 	}
+
+	statsTags := processor.StatsTagsForBatch(
+		nb.parent.name, processor.ServiceNameForNode(nb.node), nb.format,
+	)
+	_ = stats.RecordWithTags(context.Background(), statsTags, measure.M(1))
+
+	// TODO: This process should be done in an async way, perhaps with a channel + goroutine worker(s)
```

Comment:
I think the batcher doesn't need to worry about this: the next element on the pipeline takes care of that if desired.

Reply:
This measures whether the batcher is triggering a send due to a timeout or a sendSize, which can be useful when tuning how big the sendSize and timeout configurations should be

Reply:
I agree, with the metrics my point is about the TODO. We can discuss that when we either remove the TODO or implement it.

Reply:
replied to wrong line.... yea, we can see about what makes the most sense with this, it might be that we just have the other stages return back quickly and do the queuing on that side